File: BlueMS/src/main/java/com/st/BlueMS/demos/fftAmpitude/FFTAmplitudeDataStatsFragment.java
/*
 * Copyright (c) 2019  STMicroelectronics – All rights reserved
 * The STMicroelectronics corporate logo is a trademark of STMicroelectronics
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 * - Redistributions of source code must retain the above copyright notice, this list of conditions
 *   and the following disclaimer.
 *
 * - Redistributions in binary form must reproduce the above copyright notice, this list of
 *   conditions and the following disclaimer in the documentation and/or other materials provided
 *   with the distribution.
 *
 * - Neither the name nor trademarks of STMicroelectronics International N.V. nor any other
 *   STMicroelectronics company nor the names of its contributors may be used to endorse or
 *   promote products derived from this software without specific prior written permission.
 *
 * - All of the icons, pictures, logos and other images that are provided with the source code
 *   in a directory whose title begins with st_images may only be used for internal purposes and
 *   shall not be redistributed to any third party or modified in any way.
 *
 * - Any redistributions in binary form shall not include the capability to display any of the
 *   icons, pictures, logos and other images that are provided with the source code in a directory
 *   whose title begins with st_images.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
 * AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER
 * OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
 * OF SUCH DAMAGE.
 */
package com.st.BlueMS.demos.fftAmpitude;

import androidx.lifecycle.ViewModelProviders;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.st.BlueMS.R;
import com.st.BlueSTSDK.Features.FeatureMotorTimeParameter;

/**
 * dialog showing the fft statistics
 */
public class FFTAmplitudeDataStatsFragment extends DialogFragment {

    /**
     * view model containing the fft data to show
     * this fragment is shared with the plot fragment
     */
    private FFTDataViewModel mFFTViewModel;

    /**
     * view model containing the time domain data
     * this fragment is shared with the plot fragment
     */
    private TimeDomainDataViewModel mTimeDomainViewModel;

    private TextView[] mFreqStats = new TextView[3];
    private TextView mTimeStatsX;
    private TextView mTimeStatsY;
    private TextView mTimeStatsZ;

    public FFTAmplitudeDataStatsFragment(){
        super();
        setStyle(STYLE_NORMAL,R.style.DialogWithTitle);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        getDialog().setTitle(R.string.fftDetails_title);

        View root = inflater.inflate(R.layout.fragment_fft_stat, container, false);

        mFreqStats[0] = root.findViewById(R.id.fftAmpl_xData);
        mFreqStats[1] = root.findViewById(R.id.fftAmpl_yData);
        mFreqStats[2] = root.findViewById(R.id.fftAmpl_zData);

        mTimeStatsX = root.findViewById(R.id.fftAmpl_xTimeData);
        mTimeStatsY = root.findViewById(R.id.fftAmpl_yTimeData);
        mTimeStatsZ = root.findViewById(R.id.fftAmpl_zTimeData);

        return root;
    }

    private void registerFFTUpdate(){
        mFFTViewModel.getFFTMax().observe(getViewLifecycleOwner(), fftPoints -> {
            if(fftPoints==null)
                return;
            FFTComponentsConfig.LineConf[] lines = FFTComponentsConfig.LINES;

            int nComponents = Math.min(fftPoints.size(),lines.length);
            nComponents = Math.min(nComponents,mFreqStats.length);

            for (int i = 0; i <nComponents ; i++) {
                FFTDataViewModel.FFTPoint max = fftPoints.get(i);
                mFreqStats[i].setText(getString(R.string.fftDetails_freqInfo_format,
                        lines[i].name,max.amplitude,max.frequency));
            }
        });
    }

    private void updateTimeDomainLabel(TextView label, String name, @Nullable TimeDomainDataViewModel.TimeDomainStats data){
        if(data == null){
            label.setText(R.string.fftDetails_timeInfo_not_available);
        }else{
            String xData = getString(R.string.fftDetails_timeInfo_format,name,
                    data.accPeak, FeatureMotorTimeParameter.FEATURE_ACC_UNIT,
                    data.rmsSpeed,FeatureMotorTimeParameter.FEATURE_SPEED_UNIT);
            label.setText(xData);
        }
    }

    private void registerTimeDomainUpdate(){
        mTimeDomainViewModel.getXComponentStats().observe(getViewLifecycleOwner(),
                timeDomainStats -> updateTimeDomainLabel(mTimeStatsX,"X",timeDomainStats));

        mTimeDomainViewModel.getYComponentStats().observe(getViewLifecycleOwner(),
                timeDomainStats -> updateTimeDomainLabel(mTimeStatsY,"Y",timeDomainStats));

        mTimeDomainViewModel.getZComponentStats().observe(getViewLifecycleOwner(),
                timeDomainStats -> updateTimeDomainLabel(mTimeStatsZ,"Z",timeDomainStats));
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mFFTViewModel = ViewModelProviders.of(requireActivity()).get(FFTDataViewModel.class);
        mTimeDomainViewModel = ViewModelProviders.of(requireActivity()).get(TimeDomainDataViewModel.class);
        registerFFTUpdate();
        registerTimeDomainUpdate();
    }
}

File: st_fft_amplitude/src/main/java/com/st/fft_amplitude/FFTAmplitudeDataStatsFragment.kt
package com.st.fft_amplitude

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.fragment.app.DialogFragment
import androidx.hilt.navigation.fragment.hiltNavGraphViewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.st.core.ARG_NODE_ID
import com.st.fft_amplitude.databinding.FragmentFftStatBinding
import com.st.fft_amplitude.utilites.LineConf
import com.st.fft_amplitude.utilites.TimeDomainStats
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
@AndroidEntryPoint
class FFTAmplitudeDataStatsFragment : DialogFragment() {

    private val viewModel: FFTAmplitudeViewModel by hiltNavGraphViewModels(R.id.fft_amplitude_nav_graph)

    private lateinit var binding: FragmentFftStatBinding
    private lateinit var nodeId: String

    private var mFreqStats: Array<TextView?> = Array(3) {null}
    private lateinit var mTimeStatsX: TextView
    private lateinit var mTimeStatsY: TextView
    private lateinit var mTimeStatsZ: TextView

    private lateinit var mTimeStatsLinearLayout: LinearLayout
    private lateinit var mTimeStatsNotAvailable: TextView

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {

        nodeId = arguments?.getString(ARG_NODE_ID)
            ?: throw IllegalArgumentException("Missing string $ARG_NODE_ID arguments")

        binding = FragmentFftStatBinding.inflate(inflater, container, false)

        mFreqStats[0] = binding.fftAmplXData
        mFreqStats[1] = binding.fftAmplYData
        mFreqStats[2] = binding.fftAmplZData

        mTimeStatsX = binding.fftAmplXTimeData
        mTimeStatsY = binding.fftAmplYTimeData
        mTimeStatsZ = binding.fftAmplZTimeData

        mTimeStatsLinearLayout = binding.fftTimeDomainAvailable
        mTimeStatsNotAvailable = binding.fftTimeDomainNotAvailable

        return binding.root
    }

    private fun updateTimeDomainLabel(label: TextView, name: String, data: TimeDomainStats?) {
        if (data == null) {
            label.setText(R.string.fftDetails_timeInfo_not_available)
        } else {
            val xData = getString(
                R.string.fftDetails_timeInfo_format, name,
                data.accPeak, "m/s^2",
                data.rmsSpeed, "mm/s"
            )
            label.text = xData
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        //FFTUpdate
        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.mFftMax.collect {
                   val lines = LineConf.LINES
                    var nComponents = minOf(lines.size,it.size)
                    nComponents = minOf(nComponents,mFreqStats.size)

                    for (i in 0 until nComponents) {
                        val max = it[i]
                        if(mFreqStats[i]!=null) {
                            mFreqStats[i]!!.text = getString(
                                R.string.fftDetails_freqInfo_format,
                                lines[i].name, max.amplitude, max.frequency
                            )
                        }
                    }
                }
            }
        }

        //TimeDomainUpdate
        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.mXStats.collect {
                    updateTimeDomainLabel(mTimeStatsX,"X",it)
                }
            }
        }
        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.mYStats.collect {
                    updateTimeDomainLabel(mTimeStatsY,"Y",it)
                }
            }
        }
        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.mZStats.collect {
                    updateTimeDomainLabel(mTimeStatsZ,"Z",it)
                }
            }
        }

        viewModel.timeParameterAvailable.observe(this) { value ->
            if (value) {
                mTimeStatsLinearLayout.visibility = View.VISIBLE
                mTimeStatsNotAvailable.visibility = View.GONE
            } else {
                mTimeStatsLinearLayout.visibility = View.GONE
                mTimeStatsNotAvailable.visibility = View.VISIBLE
            }
        }
    }
}